catalog/core/catalog-core-subscriptionstore/subscriptionstore-impl/src/main/java/org/codice/ddf/catalog/subscriptionstore/common/CachedSubscription.java

```diff
@@ -0,0 +1,184 @@
+ * {@link #isNotRegistered()} is false.
+ * </li>
+ * </ol>
+ * Given the dynamic nature of Karaf and the need to handle {@link javax.cache.event.CacheEntryEvent}s,
+ * there may be a brief period during system startup where cached subscriptions exist in the cache but
+ * are not registered. This problem should rectify itself as additional {@link SubscriptionFactory}
+ * instances are registered by subscription endpoints.
+ * <p>
+ * Currently, no guarantees are made regarding the state of this object <i>during</i> a subscription CRUD
+ * operation. Only one container operation may be active at a time.
+ *
+ * @see org.codice.ddf.catalog.subscriptionstore.SubscriptionContainerImpl
+ */
+public class CachedSubscription {
+    private static final Logger LOGGER = LoggerFactory.getLogger(CachedSubscription.class);
+
+    private static final String SUBSCRIPTION_ID_OSGI = "subscription-id";
+
+    private static final String EVENT_ENDPOINT = "event-endpoint";
+
+    private final SubscriptionMetadata metadata;
+
+    private transient Subscription subscription;
+
+    private transient ServiceRegistration registration;
+
+    public CachedSubscription(SubscriptionMetadata metadata) {
+        notNull(metadata, "Subscription metadata cannot be null");
+        this.metadata = metadata;
+
+        this.subscription = null;
+        this.registration = null;
+    }
+
+    public SubscriptionMetadata getMetadata() {
+        return metadata;
+    }
+
+    public Optional<Subscription> getSubscription() {
+        return Optional.ofNullable(subscription);
+    }
+
+    public boolean isType(String type) {
+        return type.equals(metadata.getTypeName());
+    }
+
+    public boolean isNotType(String type) {
+        return !isType(type);
+    }
+
+    public boolean isNotRegistered() {
+        return registration == null;
+    }
+
+    /**
+     * Attempt to construct a {@link Subscription} using the provided factory. If the resultant
+     * subscription is valid, register it as this cached subscription's OSGi service.
+     *
+     * @param factory subscription factory corresponding to this cached subscription's type.
+     * @throws SubscriptionRegistrationException if the factory can't operate on this cached subscription.
+     */
+    public void registerSubscription(SubscriptionFactory factory) {
+        if (isNotType(factory.getTypeName())) {
+            throw new SubscriptionRegistrationException(format(
+                    "Factory type mismatch for subscription type [%s]",
+                    metadata.getTypeName()));
+        }
+
+        LOGGER.debug("Regenerating subscription [ {} | {} | {} ]",
+                metadata.getId(),
+                metadata.getTypeName(),
+                metadata.getCallbackAddress());
+
+        Subscription subscription = factory.createSubscription(metadata);
+        if (subscription == null) {
+            throw new SubscriptionRegistrationException("Regenerated subscription cannot be null");
```

Comment:
is this message useful to the caller?  How would they know how the regenerated subscription got to be null?  Should it be something like "Unable to regenerate subscription"?